### The First cell will always contains the libraries that i used to solve different problems so that it is organized


In [1]:
import pandas as pd
import os
import seaborn as sns
import matplotlib.pyplot as plt



### Data Exploration

In [3]:
def read_diff_files(data_dir, train_file, test_file,sample_submission):
    current_directory = os.getcwd()
    path_to_data = current_directory + data_dir  
    df_train = pd.read_csv(path_to_data + train_file)
    df_test = pd.read_csv(path_to_data+test_file)
    df_sam_subm = pd.read_csv(path_to_data+sample_submission)  
    return df_train,df_test,df_sam_subm

df_train,df_test,df_sam_subm = read_diff_files(data_dir = "/Data",train_file = "/train.csv", test_file ="/test.csv", sample_submission = "/sample_submission.csv")

### Check Null Values
1. Check the percentage of missing values and what columns 
2. See the importance of that columns 
3. Check Imputation Method
4. Do Visulatizations    

In [9]:
def data_explortation(dataframe):
    print("------------This is different dataset Info ---------------")
    print(dataframe.info())
    print("---------------------------------------------------------")
    null_counts = dataframe.isnull().sum()
    null_percentage = (null_counts/len(dataframe))* 100
    columns_with_null_percentage = null_percentage[null_percentage > 0]
    print("----------Those are the columns with null values and the percentage-------------")
    print(columns_with_null_percentage)
    print("-------------------------------------------------------------------")
    print("----------Those are the Data types in the training dataset-------------")
    data_types = list(set(dataframe.dtypes.tolist())) 
    print(data_types)
    print("-------------------------------------------------------------------")
    print("--------------This is the size of the Dataframe----------------")
    print(dataframe.shape)
    print("-----------------------------------------------------------------")

    return columns_with_null_percentage, data_types 
    # Filter columns with null values
cols_with_null_perc,date_types_df = data_explortation(df_train)    


------------This is different dataset Info ---------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object

### Check missing values and how to handle them

In [20]:
cols_with_null_perc = cols_with_null_perc.sort_values(ascending=False)
categorical_features = df_train.select_dtypes(include=['object'])
int_numeric_features = df_train.select_dtypes(include=['int64'])
float_numeric_features = df_train.select_dtypes(include=['float64'])


In [24]:
def features_types(data_frame):
    categorical_features = data_frame.select_dtypes(include=['object'])
    int_numeric_features = data_frame.select_dtypes(include=['int64'])
    float_numeric_features = data_frame.select_dtypes(include=['float64'])

    num_of_categorial_features = categorical_features.shape[1]
    num_of_numeric_features = int_numeric_features.shape[1]
    num_of_float_features = float_numeric_features.shape[1]
    
    print(f"Number of categorial features:{num_of_categorial_features} ")
    print(f"Number of numeric features:{num_of_numeric_features}")
    print(f"num_of_float_features:{num_of_float_features}")
    return categorical_features,int_numeric_features,float_numeric_features
features_types(df_train)    


Number of categorial features:43 
Number of numeric features:35
num_of_float_features:3


### Check the missing values according the feature type
 

In [29]:
def check_missing_vlaues_per_data_type(feature_type_df,datatype):
    print(f"The number of nulls in {datatype} ")
    #print(feature_type_df.isnull().sum())
    null_counts = feature_type_df.isnull().sum()
    null_percentage = (null_counts/len(feature_type_df))* 100
    columns_with_null_percentage = null_percentage[null_percentage > 0].sort_values(ascending=False)
    print(columns_with_null_percentage)

check_missing_vlaues_per_data_type(categorical_features,"categorical")
print("--------------------------------------------------------------")
check_missing_vlaues_per_data_type(int_numeric_features,"Numeric")
print("--------------------------------------------------------------")
check_missing_vlaues_per_data_type(float_numeric_features, "Float")

The number of nulls in categorical 
PoolQC          99.520548
MiscFeature     96.301370
Alley           93.767123
Fence           80.753425
FireplaceQu     47.260274
GarageType       5.547945
GarageFinish     5.547945
GarageQual       5.547945
GarageCond       5.547945
BsmtExposure     2.602740
BsmtFinType2     2.602740
BsmtQual         2.534247
BsmtCond         2.534247
BsmtFinType1     2.534247
MasVnrType       0.547945
Electrical       0.068493
dtype: float64
--------------------------------------------------------------
The number of nulls in Numeric 
Series([], dtype: float64)
--------------------------------------------------------------
The number of nulls in Float 
LotFrontage    17.739726
GarageYrBlt     5.547945
MasVnrArea      0.547945
dtype: float64


### From the above info we can conclude that
1. The most missing info is from the categorical Features, the feature with most missing values has missing vales percntage of 99.5%
2. There are no missing info from numeric features, 
3. There are missing info from float but not that big, he feature with most missing values has missing vales percntage of 17.7%